### ISSUE TITLE
Wrong session behavior when using multiple shopify shops in one browser

### ISSUE THREAD

**Issue by hpawe01** [Shopify/shopify-express#60]
There is a problem when I use the same shopify app in multiple shopify shops in one browser. The first shopify shop will set the session with `accessToken` and `shop`. But for the following shopify shops my shopify-express server thinks, that the request comes from the same (first used) shopify shop.

This issue is already reported for https://github.com/Shopify/shopify-node-app (see https://github.com/Shopify/shopify-node-app/issues/88).

The problem lies in the file https://github.com/Shopify/shopify-express/blob/master/middleware/withShop.js: It checks only, if any `accessToken` is set on the session. But not, if the `shop` that performs this request is the same as in the `session`. So we should also check, if the `shop` and the `session.shop` are identical:
```
if (session && session.accessToken && session.shop && session.shop === shop) {
```

Now we have another problem: `shop` is not always set on the `request.query`.
A possible solution could be to retrieve the `shop` either via the `query` or the `referer` header:
```
const shop = request.query.shop || request.get('referer').match(/shop=([^&]+)/)[1];
```

The result would be:
```
    // ...
    const { session } = request;
    const shop = request.query.shop || request.get('referer').match(/shop=([^&]+)/)[1];

    if (session && session.accessToken && session.shop && session.shop === shop) {
      return next();
    }
   // ...
```

Should I prepare a pull request with this change? Or does it break something I am not aware of?

**Comment by hpawe01** [comment]
One problem still exists: I cannot use two shops in parallel. When I perform a request with another shop, the oauth process of shopify is triggered again.

Maybe it is possible to store an array on the session (e.g.`authenticated shops`) that contain objects with  `accessToken` and `shop` and we verify against all elements on this array.

**Comment by TheMallen** [comment]
If you want to open a PR with that fix it looks like it could be a good solution to at least part of the problem :)

The second one is a bit hairier and could be a separate PR. I'm not sure serializing an array is the right approach, I'll think about it.

**Comment by hpawe01** [comment]
Done. And I am really interested about your thoughts on the second problem.

**Comment by TheMallen** [comment]
I *think* we can solve the second problem while solving the new issue I just opened about security. I'll probably try it out in the next day or two. :)

**Comment by paulisloud** [comment]
Done some testing on this in the last couple of days. Based on all the various scenarios (reload app, login, browse a route, install app from app store, install app via hosted install route), this was was the biz for me:

    module.exports = function withShop({ authBaseUrl } = {}) {
      return function verifyRequest(request, response, next) {
        let { query: { shop }, session = {}, baseUrl} = request;

        if (shop === undefined && session.shop) {
          shop = session.shop
        }

        if (!shop && request.get('referer')) {
          shop = shopFromReferrer(request.get('referrer')) || session.shop;
        }

        if (session && session.accessToken && session.shop && session.shop === shop) {
          next();
          return;
        }

        if (shop) {
          response.redirect(`${authBaseUrl || baseUrl}/auth?shop=${shop}`);
          return;
        }

        response.redirect('/install');
        return;
      };
    };


Then, defining `shopFromReferrer` separately, as I believe @TheMallen suggested:

    function shopFromReferrer(referrer) {
      const results = referrer.match(/shop=([^&]+)/) ;
      return results && results[1]
    }

Does this bring up any security issues?

**Comment by SeanoNET** [comment]
Hi, I am having the same issues, is a fix scheduled for release? Or since this library is now deprecated this will no longer be patched?
